### PR TITLE
WIP Migrations

### PR DESCRIPTION
**Pull request by slifty**
We're moving our database setup to Liquibase migrations.  This PR contains the initial setup of that change as well an initial changeset just to get things started.  In future each ChangeSet will get its own commit; this was a special case because the files didn't exist yet, so there were no changes to patch into them. The bulk of the migration scripts themselves will be separate PRs.

Two other important notes:

1) We had talked about having the liquibase tasks be part of an existing project; this made sense initially, but our directory structure already places `db` as its own project (which is also best practice). By mirroring this structure in gradle we get much more intuitive migration commands (e.g. `db:update` instead of `service:update`), which feels much more natural to me.

I feel like this is using the concept of projects appropriately -- and that the database is an encapsulated, independent component of the system (basically, it is reasonable to organize its management logic as such.)

2) I spent a little bit of time looking for the right place to put automated liquibase code as part of service startup, I decided not to go through with it for this PR.  I feel that the gradle approach gives full control while still script-able and therefore potentially automated as part of deployment scripts.  Embedding migrations into code, on the other hand, would remove agency from developers and administrators who might not want migrations to be run outside of formal deployments.

In other words: whenever an updated version of the application is deployed it should be trivial to include a line to the deployment script to run the migrations (whereas it would not be trivial to have migrations NOT run when the application is started, if calls to migrations were hard coded into service).

Happy to discuss either point in more detail.

---

Issue #254 Allow database schema changes without losing data

### CODE REVIEW

File: psm-app/build.gradle [hunk 1]
```diff
@@ -6,6 +6,7 @@ buildscript {
         maven { url 'https://plugins.gradle.org/m2/' }
     }
     dependencies {
+        classpath 'org.liquibase:liquibase-gradle-plugin:1.2.0'
```
Comment (jasonaowen) [inline psm-app/build.gradle]: It looks like there's [a newer version](https://search.maven.org/#artifactdetails%7Corg.liquibase%7Cliquibase-gradle-plugin%7C1.2.4%7Cjar); can we use it, or do we need to stay on `1.2.0`?

File: psm-app/build.gradle [hunk 2]
```diff
@@ -6,12 +6,14 @@ buildscript {
...
     }
     dependencies {
+        classpath 'org.liquibase:liquibase-gradle-plugin:1.2.4'
         classpath 'gradle.plugin.org.openrepose:gradle-jaxb-plugin:2.4.1'
         classpath 'com.moowork.gradle:gradle-node-plugin:1.2.0'
         classpath 'kr.motd.gradle:sphinx-gradle-plugin:1.0.5'
         classpath ('net.serenity-bdd:serenity-gradle-plugin:1.8.20') {
             exclude group: 'com.google.inject', module:'guice'
         }
+        classpath System.properties['databaseDriver']
```
Comment (jasonaowen) [inline psm-app/build.gradle]: I think we need to have either a default value, or an explicit check for this value being present so that we can display a more useful error. Right now, CI (which doesn't have a `gradle.properties`) is complaining:

> Cannot convert a null value to an object of type Dependency.

Alternatively, I think it would be okay to hard-code a dependency to PostgreSQL here; if we ever need to change databases, we'll need to change the Hibernate driver, as well, which is also currently hard-coded.

File: psm-app/gradle.properties.template [hunk 3]
```diff
@@ -0,0 +1,4 @@
+systemProp.databaseUser=
+systemProp.databasePassword=
+systemProp.databasePath=jdbc:postgresql:psm
+systemProp.databaseDriver=postgresql:postgresql:9.1-901-1.jdbc4
```
Comment (jasonaowen) [inline psm-app/gradle.properties.template]: This is a super old driver; we should use something more recent, like [`org.postgresql:postgresql:42.2.2`](https://search.maven.org/#artifactdetails%7Corg.postgresql%7Cpostgresql%7C42.2.2%7Cbundle).

File: docs/MIGRATIONS.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+```shell
+> ./gradlew db:update
+```
+
+## Adding Migrations
+
+Future changes to the database are performed by adding ChangeSets to the ChangeLog associated with the current named version.
+
+ChangeSets can never be updated or deleted once committed to master.  If a past ChangeSet needs to be modified, create a new Changeset with the appropriate alteration commands.
```
Comment (jasonaowen) [inline docs/MIGRATIONS.md]: Can you please add why this is? Preferably with a link to the docs so we can read more.

File: docs/MIGRATIONS.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# MIGRATIONS
+The PSM uses [Liquibase](http://liquibase.org/) to manage migrations. Review the [Liquibase best practices](http://www.liquibase.org/bestpractices.html) to understand the structure of this project's migrations.
+
+Migrations are stored in a `changelog` folder within the `psm-app/db/` project.  The `db-changelog-master.xml` is the root changelog.
+
+Changes to the database are registered inside of `ChangeLog` files. Each change in a ChangeLog is called a `ChangeSet`. ChangeSets should be as atomic as possible, and should also contain `rollback` instructions which allows them to be undone.
+
+## Running Migrations
+To run migrations
```
Comment (jasonaowen) [inline docs/MIGRATIONS.md]: How do you migrate down?

Comment (jasonaowen) [reply docs/MIGRATIONS.md]: Actually, maybe it'd be most useful to document how to get help! `./gradlew db:tasks` has super useful output, including such highlights as:

```
Liquibase tasks
---------------
changelogSync - Mark all changes as executed in the database.
changelogSyncSQL - Writes SQL to mark all changes as executed in the database to STDOUT.
markNextChangesetRan - Mark the next change set as executed in the database.
rollback - Rolls back the database to the state it was in when the <liquibaseCommandValue> tag was applied.
rollbackToDate - Rolls back the database to the state it was in at the <liquibaseCommandValue> date/time.
update - Updates the database to the current version.
updateSQL - Writes SQL to update the database to the current version to STDOUT.
validate - Checks the changelog for errors.
```

Although I'm not sure what `<liquibaseCommandValue>` is; do you know, @slifty?

**Comment by slifty** [comment]
Jenkins retest this please

**Comment by jasonaowen** [comment]
Ah good! These test failures look like what we'd expect from not having run the migrations.

I'll look at how to set up Jenkins in a way that won't break the other pull requests that don't yet have this branch merged.

File: docs/MIGRATIONS.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+ChangeSets should never be updated or deleted once committed to master.  If a past ChangeSet needs to be modified, create a new Changeset with the appropriate alteration commands.
+
+The "always append changes" rule is enforced by Liquibase in the form of checksums for ChangeSets -- if you edit a ChangeSet that has already been run, Liquibase will complain loudly.
+
+The reasoning for this [is explained on this liquibase blog post](http://www.liquibase.org/2008/10/dealing-with-changing-changesets.html).  That post also explains how to bypass the Liquibase safeguards in the event of a big mistake that truly requires a rewriting of history.
+
+```
+The goal of a Liquibase change log file is to track the linear sequence of changes required to take a database from a starting point to the current state, and it is built up one change at a time throughtout development. By following the “always append changes” rule you will ensure that all databases are consistent since all will have gone through the exact same set of changes.
```
Comment (jasonaowen) [inline docs/MIGRATIONS.md]: Would you mind making this a blockquote instead of a code section? That way readers don't have to horizontally scroll in the tiny code view window.

File: docs/MIGRATIONS.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+## Adding Migrations
+
+Future changes to the database are performed by adding ChangeSets to the ChangeLog associated with the current named version.  For instance `db/changelog/db-changelog-1.0.xml`.
+
+ChangeSets should never be updated or deleted once committed to master.  If a past ChangeSet needs to be modified, create a new Changeset with the appropriate alteration commands.
+
+The "always append changes" rule is enforced by Liquibase in the form of checksums for ChangeSets -- if you edit a ChangeSet that has already been run, Liquibase will complain loudly.
+
+The reasoning for this [is explained on this liquibase blog post](http://www.liquibase.org/2008/10/dealing-with-changing-changesets.html).  That post also explains how to bypass the Liquibase safeguards in the event of a big mistake that truly requires a rewriting of history.
```
Comment (jasonaowen) [inline docs/MIGRATIONS.md]: This is great, thank you!

**Comment by jasonaowen** [comment]
After some further consideration, I think we need some more documentation. Much of the content in `docs/MIGRATIONS.md` might be better in `psm-app/db/README.md`, and some should be put in `INSTALL.md` - in particular, we should update [the section about deploying](https://github.com/SolutionGuidance/psm/blob/6abec5a8d310d0af0ff6ca1f3202faa03fe274b8/INSTALL.md#build-and-deploy-the-application).

What do you think, @slifty?

**Comment by slifty** [comment]
@jasonaowen absolutely, however I'd really like to merge this as it stands so that I can start to make progress on the migrations.

The focus of this PR wasn't ever meant to be the documentation -- I mostly was creating a placeholder with some basics which we could iterate on separately.

**Comment by jasonaowen** [comment]
> The focus of this PR wasn't ever meant to be the documentation

I had forgotten that we were going to defer the mailing list announcement until after the convert-to-migrations sequence of PRs was finished, and so I was thinking that we needed the documentation ready so that we could point people to it and say "hey, this is how you deploy now". I think we agree that that needs to happen before we're done, but now I realize that it doesn't need to happen in this pull request - I'm sorry for my confusion!

**Comment by jasonaowen** [comment]
The integration tests passed, and I was able to run the `changelogSync` task to mark my database as up-to-date without resetting it:

```psql
psm=> select * from databasechangelog;
-[ RECORD 1 ]-+----------------------------------------------
id            | issue254-create-hibernate-sequence
author        | 
filename      | db/changelog/db-changelog-1.0.xml
dateexecuted  | 2018-05-03 14:06:40.38419
orderexecuted | 1
exectype      | EXECUTED
md5sum        | 7:afe897bea20bc08d1681087c94d0967d
description   | createSequence
comments      | 
tag           | 
liquibase     | 3.4.2
contexts      | 
labels        | 
-[ RECORD 2 ]-+----------------------------------------------
id            | issue254-create-table-persistent-logins-table
author        | 
filename      | db/changelog/db-changelog-1.0.xml
dateexecuted  | 2018-05-03 14:06:40.393397
orderexecuted | 2
exectype      | EXECUTED
md5sum        | 7:5ca34ad35b7d437df798d47a045cd5b1
description   | createTable
comments      | 
tag           | 
liquibase     | 3.4.2
contexts      | 
labels        | 
```

I think this is ready! Thanks, @slifty :tada:

**Comment by jasonaowen** [comment]
I've updated #254 with the list of things that I think needs to be done to resolve it; here's a copy of what I put in that issue, plus a bit of commentary:

- Decide on migration strategy (done! it's liquibase)
- Allow manual migrations (done when this PR is merged! :tada: )
- Document how to run manual migrations (partially done, more to do)
- Convert contents of `seed.sql` and `jbpm.sql` to new migration language (partially done, and the subject of several subsequent PRs)
- Automatically run migrations in our CD environment
- Announce resolution of this issue on the mailing list

@slifty, when you merge this PR, will you check off the second item on the list over there? Also, if there's anything I've forgotten, please add it to that checklist!

**Comment by slifty** [comment]
Thank you @jasonaowen and @cecilia-donnelly!  This all looks great, (and no problem around the various requests; I'll also start a parallel branch for improving the documentation based on the feedback provided here while working on the migration PRs)
